bacula/manpages/bconsole.8
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH BCONSOLE 8 "4 December 2009" "Kern Sibbald" "Network backup, recovery and verification"
.\" Please adjust this date whenever revising the manpage.
.\"
.SH NAME
 bconsole \- Bacula's management Console
.SH SYNOPSIS
.B bconsole
.RI [options]
.br
.SH DESCRIPTION
This manual page documents briefly the
.B bconsole
command.
.PP
.SH OPTIONS
.TP
.BI \-D\  dir
Select a Director.
.TP
.BI \-l
List defined Directors.
.TP
.BI \-c\  config
Specify configuration file.
.TP
.BI \-d\  nn
Set debug level to \fInn\fP.
.TP
.BI \-dt
Print timestamp in debug output.
.TP
.B \-n
No conio (for scripting).
.TP
.B \-s
No signals (for debugging).
.TP
.b \-u\  nn
Set command execution timeout to \fInn\fP seconds.
.TP
.B \-t
Test the configuration file and report errors.
.TP
.B \-?
Show version and usage of program.
.SH SEE ALSO
.BR bacula\-dir (8),
.BR bls (1),
.BR bextract (1).
.br
.SH AUTHOR
This manual page was written by Jose Luis Tallon
.nh 
<jltallon@adv\-solutions.net>.
.SH COPYRIGHT
This man page document is released under the BSD 2-Clause license.


bacula/ChangeLog
              Changelog on version 7.0.5

Release version 7.0.5

28Jul14
 - Fix #547 by adding .schedule command

27Jul14
 - Update AUTHORS
 - Fix bug #2079 with patch from Robert Oschwald

26Jul14
 - Fix orphaned file descriptors during errors
 - Yet another client==NULL
 - Improve FD and SD cancel
 - Jim Raney's TLS patch
 - Update AUTHORS
 - Fix bug #1679 pool overrides not shown in manual run display
 - Attempt to avoid client==NULL

23Jul14
 - Fix for bug #2082 (hopefully)

22Jul14
 - Fix seg fault in jobq.c

14Jul14
 - make stop after first error

12Jul14
 - Increase status schedule days from 500 to 3000
 - Remove bad cherry-pick
 - Fix compiler warning
 - Allow options create_postgresql_database from patch in bug #2075 by roos
 - Fix bug #2074 crashes when no conf file present

10Jul14
 - Set pthread id in jcr at beginning so the job can be canceled.
 - Fix possible heartbeat interval timing problems

08Jul14
 - Fix some errors reported by valgrind. May fix the problem with bsmtp command.
 - Ensure b_sterror() is using the correct size of input buffer

07Jul14
 - Fix possible seg fault

04Jul14
 - Fix segfault when trying to stop the bnet_server thread in terminate_stored()

03Jul14
 - Fix bad link bug #2076

02Jul14
 - Fix compilation of bsock.c when TLS is not available

27Jun14
 - Correct L suffix to be LL
 - Fix bad copy/migrate data header

26Jun14
 - On termination shutdown thread server

23Jun14
 - baculum: Updated README file
 - baculum: Update English language texts
 - baculum: Saving auth file for web server HTTP Basic auth
 - baculum: Added directory for web server logs
 - baculum: Added example Lighttpd configuration for Baculum and sample web 
  server auth file
 - Expanded auth error message
 - baculum: Support for web servers which do not provide direct info about HTTP 
  Basic auth

15Jun14
 - Fix limit bandwidth calculation
 - Eliminate strcpy() from bsmtp

12Jun14
 - Fix for configuring sudo option for bconsole access
 - Display correct NextPool overrides + use Job NextPool in restore if available

09Jun14
 - Fix Bacula to work with newer Windows pthreads library

17May14
 - Fix bug #180 ERR=success in btape when tape error

Bugs fixed/closed since last release:
1679 180 2074 2075 2076 2079 2082 547

====

Release version 7.0.4
02Jun14
 - Better error handling for cancel command

01Jun14
 - Fix compiler warning + simplify some #ifdefs

22May14
 - Fix copy/migration to second SD

19May14
 - Fix calls to sl.set_string()
 - Improve sellist code

=====

Release version 7.0.3
12May14
 - Fix error handling in do_alist_prompt
 - Tighten error condition handling in sellist
 - Add new cancel test

06May14
 - Update LICENSE and LICENSE-FAQ

03May14
 - Also update autoconf/aclocal.m4
 - Reschedule on error caused EndTime to be incorrect -- fixes bug #2029
 - Flush console queued job messages -- should fix bug #2054
 - Attempt to fix FreeBSD echo/printf, bug #2048
 - Update to newer libtool + config.guess
 - Recompile configure
 - Apply fix supplied for acl.c in bug #2050

01May14
 - Fix a SD seg fault that occurs with over committed drives

28Apr14
 - Clear bvfs cache and set debug options available only for admin
 - Moved auth params to curl opts
 - Filtred single results for restricted consoles

27Apr14
 - Removed unnecessary debug
 - Changed e-mail address in gettext file
 - Support for customized and restricted consoles

15Apr14
 - Misc changes for rpm building (made by Louis)

13Apr14
 - Updated requirements for Baculum

12Apr14
 - Apply fix for bug 2049: wrong drive selected

11Apr14
 - Fix #2047 about bthread_cond_wait_p not declared

09Apr14
 - Fix Bacula bug #2044 -- fix Makefile for bplugininfo linking
 - Fix Bacula bug #2046 -- sellist limited to 10000
 - Fix Bacula bug #2045 -- multiply defined daemon_event
 - Fix Bacula bug #2020 overflow in btape -- Andreas Koch

Bugs fixed/closed since last release:
2020 2029 2044 2045 2046 2047 2048 2050 2054


Release version 7.0.2
02Apr14
 - Remove more vestiges of libbacpy
 - Put back @PYTHON@ path in configure
 - Fix improper string in parser
 - Remove libbacpy from rpm spec files
 - Fix linking check_bacula
 - Fix new SD login in check_bacula
 - Tweek docs build process

Release version 7.0.1
31Mar14
 - Remove old plugin-test
 - Update po files
 - Enable installation of the bpluginfo utility
 - More tray-monitor updates
 - Add Simone Caronii to AUTHORS
 - Align command line switches in manpages.
 - Apply upgrade to config.guess
 - Remove bgnome-console and bwx-console leftovers.
 - Update tray-monitor header also for new bsock calls
 - Attempt to fix nagios to use new bsock calls
 - Update tray-monitor to new bsock calls

==========
Release version 7.0.0

24Mar14
 - Add Josip Almasi to AUTHORS
 - [PATCH] Support for restricted consoles in BAT config
 - [PATCH] Fix for free director directive
 - [PATCH] Fix auto-select restricted console for director in bconsole
 - Realign output display
 - Update ua_output.c from Branch-6.7
 - Add some missing Branch-6.7 updates
 - Added needed empty directories to Baculum
 - Fix for support PostgreSQL, MySQL and SQLite
 - Framework adjusting to Baculum database connections
 - Framework fix for lower case tables names in MySQL
 - Fix for Baculum SQLite support
 - Initial commit Baculum
 - Add Marcin to AUTHORS file
 - Strip trailing blanks
 - Update copyright year
 - Update LICENSE and header files
 - Remove old file
 - Add new header in misc files
 - Remove tray-monitor bwx-console manual installation
 - Remove FD python and examples
 - Fixup spec files
 - Remove pythonlib from lib
 - Update package-list
 - Fix SDCallsClient daemon synchronization
 - Add debug code + make 127.0.0.1 same as localhost for tls tests
 - Fix multiple DIRs in console
 - Make failure for bat to connect to DIR non-fatal
 - Fix bat style to one that works
 - Take disk-changer from Branch-6.7
 - Simplify Version output
 - Fix FDVersion for SD Calls Client test
 - Update accurate test
 - Update differential test

gui/baculum/protected/Class/Bconsole.php
<?php
/**
 * Bacula® - The Network Backup Solution
 * Baculum - Bacula web interface
 *
 * Copyright (C) 2013-2015 Marcin Haba
 *
 * The main author of Baculum is Marcin Haba.
 * The main author of Bacula is Kern Sibbald, with contributions from many
 * others, a complete list can be found in the file AUTHORS.
 *
 * You may use this file and others of this release according to the
 * license defined in the LICENSE file, which includes the Affero General
 * Public License, v3.0 ("AGPLv3") and some additional permissions and
 * terms pursuant to its AGPLv3 Section 7.
 *
 * Bacula® is a registered trademark of Kern Sibbald.
 */

Prado::using('Application.Class.ConfigurationManager');
Prado::using('Application.Class.Logging');
Prado::using('Application.Class.Errors');

class Bconsole extends TModule {

	const SUDO = 'sudo';

	const BCONSOLE_COMMAND_PATTERN = "%s%s -c %s %s <<END_OF_DATA\n%s\nquit\n<<END_OF_DATA";

	const BCONSOLE_DIRECTORS_PATTERN = "%s%s -c %s -l";

	const BCONSOLE_CFG_USER_KEYWORD = '{user}';

	private $availableCommands = array('version', 'status', 'list', 'messages', 'show', 'mount', 'umount', 'release', 'prune', 'purge', 'update', 'estimate', 'run', '.bvfs_update', '.bvfs_lsdirs', '.bvfs_lsfiles', '.bvfs_versions', '.bvfs_get_jobids', '.bvfs_restore', '.bvfs_clear_cache', 'restore', 'cancel', 'delete', '.jobs', 'label', 'reload', '.fileset', '.storage', '.client', '.pool');

	private $useSudo = false;

	private $bconsoleCmdPath;

	private $bconsoleCfgPath;

	private $bconsoleCfgCustomPath;

	public function init($config) {
		if($this->Application->getModule('configuration')->isApplicationConfig() === true) {
			$params = ConfigurationManager::getApplicationConfig();
			$useSudo = ((integer)$params['bconsole']['use_sudo'] === 1);
			$bconsoleCmdPath = $params['bconsole']['bin_path'];
			$bconsoleCfgPath = $params['bconsole']['cfg_path'];
			$bconsoleCfgCustomPath = array_key_exists('cfg_custom_path', $params['bconsole']) ? $params['bconsole']['cfg_custom_path'] : null;
			$this->setEnvironmentParams($bconsoleCmdPath, $bconsoleCfgPath, $bconsoleCfgCustomPath, $useSudo);
		}
	}

	private function setEnvironmentParams($bconsoleCmdPath, $bconsoleCfgPath, $bconsoleCfgCustomPath, $useSudo) {
		$this->bconsoleCmdPath = $bconsoleCmdPath;
		$this->bconsoleCfgPath = $bconsoleCfgPath;
		$this->bconsoleCfgCustomPath = $bconsoleCfgCustomPath;
		$this->useSudo = $useSudo;
	}

	private function isCommandValid($command) {
		$command = trim($command);
		return in_array($command, $this->availableCommands);
	}

	private function prepareResult(array $output, $exitcode, $bconsoleCommand) {
		array_pop($output); // deleted 'quit' bconsole command
		for($i = 0; $i < count($output); $i++) {
			if(strstr($output[$i], $bconsoleCommand) == false) {
				unset($output[$i]);
			} else {
				break;
			}
		}
		$output = count($output) > 1 ? array_values($output) : array_shift($output);
		return (object)array('output' => $output, 'exitcode' => $exitcode);
	}

	public function bconsoleCommand($director, array $command, $user = null) {
		$baseCommand = count($command) > 0 ? $command[0] : null;
		if($this->isCommandValid($baseCommand) === true) {
			$result = $this->execCommand($director, $command, $user);
		} else {
			$result = $this->prepareResult(array(BconsoleError::MSG_ERROR_INVALID_COMMAND, ''), BconsoleError::ERROR_INVALID_COMMAND, ' ');
		}
		return $result;
	}

	private function execCommand($director, array $command, $user) {
		if(!is_null($director) && $this->isValidDirector($director) === false) {
			$output = array(BconsoleError::MSG_ERROR_INVALID_DIRECTOR, '');
			$exitcode = BconsoleError::ERROR_INVALID_DIRECTOR;
			$result = $this->prepareResult($output, $exitcode, ' ');
		} else {
			$dir = is_null($director) ? '': '-D ' . $director;
			$sudo = ($this->useSudo === true) ? self::SUDO . ' ' : '';
			$bconsoleCommand = implode(' ', $command);
			if(!is_null($this->bconsoleCfgCustomPath) && !is_null($user)) {
				$this->bconsoleCfgPath = str_replace(self::BCONSOLE_CFG_USER_KEYWORD, $user, $this->bconsoleCfgCustomPath);
			}
			$cmd = sprintf(self::BCONSOLE_COMMAND_PATTERN, $sudo, $this->bconsoleCmdPath, $this->bconsoleCfgPath, $dir, $bconsoleCommand);
			exec($cmd, $output, $exitcode);
			if($exitcode != 0) {
				$output = array(BconsoleError::MSG_ERROR_BCONSOLE_CONNECTION_PROBLEM, '');
				$exitcode = BconsoleError::ERROR_BCONSOLE_CONNECTION_PROBLEM;
				$result = $this->prepareResult($output, $exitcode, ' ');
			} else {
				$result = $this->prepareResult($output, $exitcode, $bconsoleCommand);
			}
		}
		$this->Application->getModule('logging')->log($cmd, $result, Logging::CATEGORY_EXECUTE, __FILE__, __LINE__);

		return $result;
	}

	public function getDirectors() {
		$sudo = ($this->useSudo === true) ? self::SUDO . ' ' : '';
		$cmd = sprintf(self::BCONSOLE_DIRECTORS_PATTERN, $sudo, $this->bconsoleCmdPath, $this->bconsoleCfgPath);
		exec($cmd, $output, $exitcode);
		if($exitcode != 0) {
			$output = array(BconsoleError::MSG_ERROR_BCONSOLE_CONNECTION_PROBLEM, '');
			$exitcode = BconsoleError::ERROR_BCONSOLE_CONNECTION_PROBLEM;
		}
		$result = (object)array('output' => $output, 'exitcode' => $exitcode);
		return $result;
	}

	private function isValidDirector($director) {
		$ret = in_array($director, $this->getDirectors()->output);
		return $ret;
	}

	public function testBconsoleCommand(array $command, $bconsoleCmdPath, $bconsoleCfgPath, $useSudo) {
		$this->setEnvironmentParams($bconsoleCmdPath, $bconsoleCfgPath, null, $useSudo);
		$director = array_shift($this->getDirectors()->output);
		return $this->bconsoleCommand($director, $command);		
	}
}
?>